src/storage/buffer_manager/bm_file_handle.cpp

```diff
@@ -62,124 +67,124 @@ void BMFileHandle::addNewPageGroupWithoutLock() {
-    // Note that we do not have to acquire an xlock here because this function assumes that prior to
-    // calling this function,  pageVersion and pageGroupLocks have been resized correctly.
-    std::shared_lock sLck{fhSharedMutex};
-    assert(pageIdx < numPages);
+    std::unique_lock xLck{fhSharedMutex};
+    assert(originalPageIdx < numPages);
     // Although getPageElementCursorForPos is written to get offsets of elements
     // in pages, it simply can be used to find the group/chunk and offset/pos in group/chunk for
     // any chunked data structure.
-    auto pageGroupIdxAndPosInGroup =
-        PageUtils::getPageElementCursorForPos(pageIdx, StorageConstants::PAGE_GROUP_SIZE);
-    // If we have not created a vector of locks and pageVersion array for each page in this
-    // group, first create them.
-    bool pageGroupLockAcquired = false;
-    while (!pageGroupLockAcquired) { // spinning wait
-        pageGroupLockAcquired = !pageGroupLocks[pageGroupIdxAndPosInGroup.pageIdx]->test_and_set();
-    }
-    if (pageVersions[pageGroupIdxAndPosInGroup.pageIdx].empty()) {
-        pageVersions[pageGroupIdxAndPosInGroup.pageIdx].resize(
-            StorageConstants::PAGE_GROUP_SIZE, UINT32_MAX);
+    auto [pageGroupIdx, _] =
+        PageUtils::getPageElementCursorForPos(originalPageIdx, StorageConstants::PAGE_GROUP_SIZE);
+    if (!walPageIdxGroups.contains(pageGroupIdx)) {
+        walPageIdxGroups.insert(std::make_pair(pageGroupIdx, std::make_unique<WALPageIdxGroup>()));
     }
-    pageGroupLocks[pageGroupIdxAndPosInGroup.pageIdx]->clear();
+    return pageGroupIdx;
 }
 
 void BMFileHandle::resetToZeroPagesAndPageCapacity() {
-    std::unique_lock xlock(fhSharedMutex);
-    numPages = 0;
-    pageCapacity = 0;
+    removePageIdxAndTruncateIfNecessary(0 /* pageIdx */);
     FileUtils::truncateFileToEmpty(fileInfo.get());
-    initPageStatesAndGroups();
 }
 
 void BMFileHandle::removePageIdxAndTruncateIfNecessary(common::page_idx_t pageIdx) {
     std::unique_lock xLck{fhSharedMutex};
-    removePageIdxAndTruncateIfNecessaryWithoutLock(pageIdx);
-}
-
-void BMFileHandle::removePageIdxAndTruncateIfNecessaryWithoutLock(
-    common::page_idx_t pageIdxToRemove) {
-    if (numPages <= pageIdxToRemove) {
+    if (numPages <= pageIdx) {
         return;
     }
-    for (auto pageIdx = pageIdxToRemove; pageIdx < numPages; ++pageIdx) {
-        pageStates[pageIdx].reset();
-    }
-    numPages = pageIdxToRemove;
+    numPages = pageIdx;
+    pageStates.resize(numPages);
     auto numPageGroups = getNumPageGroups();
     if (numPageGroups == frameGroupIdxes.size()) {
         return;
     }
     assert(numPageGroups < frameGroupIdxes.size());
     frameGroupIdxes.resize(numPageGroups);
-    pageGroupLocks.resize(numPageGroups);
-    pageVersions.resize(numPageGroups);
+    for (auto& [pageGroupIdx, _] : walPageIdxGroups) {
+        if (pageGroupIdx >= numPageGroups) {
+            walPageIdxGroups.erase(pageGroupIdx);
```

Comment:
I wonder if you should also check if walPageIdxGroups has become empty() and if so clear() or reconstruct it. My worry is that this map is getting inserted things and erased things continuously but we don't run any code to clear() it at some point. Not sure how badly maps can behave if they are continuously inserted into and emptied.

Reply:
Changed the logic to not erase while iterating, and directly clear when numPageGroups is 0.